### PR TITLE
CA-289145: Close socket if error occurs when connecting

### PR DESCRIPTION
**Pull request by krizex**
Signed-off-by: Yang Qian <yang.qian@citrix.com>

### CODE REVIEW

File: src/io.ml [hunk 1]
```diff
@@ -37,8 +37,12 @@ let write fd buf =
 let connect host port =
 	let sockaddr = Unix.ADDR_INET (host, port) in
 	let fd = Unix.socket Unix.PF_INET Unix.SOCK_STREAM 0 in
-	Unix.connect fd sockaddr;
-	fd
+	try
+		Unix.connect fd sockaddr;
+		fd
+	with e ->
+		Unix.close fd;
```
Comment (mseri) [inline src/io.ml]: There is a lot of duplication with https://github.com/xapi-project/xenopsd/pull/512/files#diff-8ebf9b785c8074bebb779899dfa573a6

Why can't we use this io module in xenopsd instead? What is the difference between the two modules?

Comment (krizex) [reply src/io.ml]: Why there is a repo `xenopsd` and another `xenops`? Is `xenops` been used anywhere?

Comment (mseri) [reply src/io.ml]: There is https://github.com/xapi-project/xenops that publishes xapi
```
$ opam list --depends-on xapi-xenops
# Available packages directly depending on xapi-xenops.master for 4.05.0:
xapi-rrdd          master  Performance monitoring daemon for xapi
xapi-xenops-utils  master  Create/destroy/manipulate Xen domains
-xenops and xapi-xenops-utils.
```
If there is overlapping with xenopsd, unless we have a good reason I'd reduce duplication and use xenops here as well. I'd like to hear an opinion about this from @robhoes or @jonludlam that know the history of these packages

File: .travis.yml [hunk 2]
```diff
@@ -5,9 +5,9 @@ before_install:
     - sudo add-apt-repository -y "deb http://archive.ubuntu.com/ubuntu/ trusty main universe"
     - sudo apt-get update -qq
     - sudo apt-get install libgcrypt20 liblzma5 libselinux1
-    - wget "http://archive.ubuntu.com/ubuntu/ubuntu/pool/main/s/systemd/libsystemd0_219-7ubuntu3_amd64.deb"
-    - wget "http://archive.ubuntu.com/ubuntu/ubuntu/pool/main/s/systemd/libsystemd-dev_219-7ubuntu3_amd64.deb"
-    - sudo dpkg -i libsystemd0_219-7ubuntu3_amd64.deb libsystemd-dev_219-7ubuntu3_amd64.deb
```
Comment (mseri) [inline .travis.yml]: We should probably move to travis-opam, see https://github.com/xapi-project/wsproxy/commit/c501300d4aa3f37011fac62c4ca256d0f17a1fa4

File: src/io.ml (hunk 1 above)
Comment (robhoes) [reply src/io.ml]: Either that, or move whatever is left in here that is useful (mostly `list_domains`) into xenopsd.git, and remove this repo. IMO, this one was a disaggregation step to far... @jonludlam ?

Comment (krizex) [reply src/io.ml]: For the two repos depends on `xapi-xenops`, I found:
1. `xapi-rrdd` could build without installing `xapi-xenops`
2. no package depens on `xapi-xenops-utils`

Does that means we could remove/archive this repo?

Comment (mseri) [reply src/io.ml]: Sounds like it. I think the `list_domains` binary (part of the utils) is installed as part of xenserver, so we would need to preserve and install that as part of xenopsd

**Comment by mseri** [comment]
Please add a ticket to the toolstack backlog to cleanup `xapi-xenops` as suggested above by @robhoes

File: .travis.yml [hunk 3]
```diff
@@ -5,9 +5,9 @@ before_install:
...
+    - wget "http://archive.ubuntu.com/ubuntu/ubuntu/pool/main/s/systemd/libsystemd0_229-4ubuntu4_amd64.deb"
+    - wget "http://archive.ubuntu.com/ubuntu/ubuntu/pool/main/s/systemd/libsystemd-dev_229-4ubuntu4_amd64.deb"
+    - sudo dpkg -i libsystemd0_229-4ubuntu4_amd64.deb libsystemd-dev_229-4ubuntu4_amd64.deb
     - git clone https://github.com/xapi-project/xs-opam /tmp/xs-opam && cd /tmp/xs-opam && sh into_repo.sh && cd -
```
Comment (mseri) [inline .travis.yml]: This has to call `sh tools/into_repo.sh`

**Comment by krizex** [comment]
Created CP-28272

**Pull request merged**